acceptance/detector_test.go

```diff
@@ -0,0 +1,209 @@
+	"runtime"
+	"strings"
+	"testing"
+	"time"
+
+	h "github.com/buildpacks/lifecycle/testhelpers"
+
+	"github.com/sclevine/spec"
+	"github.com/sclevine/spec/report"
+)
+
+var (
+	detectDockerContext = filepath.Join("testdata", "detector")
+	detectorBinaryDir    = filepath.Join("testdata", "detector", "container", "cnb", "lifecycle")
+	detectImage         = "lifecycle/acceptance/detector"
+	userID = "1234"
+)
+
+func TestDetector(t *testing.T) {
+	h.SkipIf(t, runtime.GOOS == "windows", "Detector is not yet supported on Windows")
+
+	rand.Seed(time.Now().UTC().UnixNano())
+
+	h.MakeAndCopyLifecycle(t, "linux", detectorBinaryDir)
+	h.DockerBuild(t, detectImage, detectDockerContext)
+	defer h.DockerImageRemove(t, detectImage)
+
+	spec.Run(t, "acceptance-detector", testDetector, spec.Parallel(), spec.Report(report.Terminal{}))
+}
+
+func testDetector(t *testing.T, when spec.G, it spec.S) {
+	when("called with arguments", func() {
+		it("errors", func() {
+			cmd := exec.Command("docker", "run", "--rm", detectImage, "some-arg")
+			output, err := cmd.CombinedOutput()
+			h.AssertNotNil(t, err)
+			expected := "failed to parse arguments: received unexpected arguments"
+			h.AssertStringContains(t, string(output), expected)
+		})
+	})
+
+	when("running as a root", func() {
+		it("errors", func() {
+			cmd := exec.Command("docker", "run", "--rm", detectImage)
+			output, err := cmd.CombinedOutput()
+			h.AssertNotNil(t, err)
+			expected := "failed to build: refusing to run as root"
+			h.AssertStringContains(t, string(output), expected)
+		})
+	})
+
+	when("read buildpack order file failed", func() {
+		it("errors", func() {
+			// no order.toml file in the default directory
+			cmd := exec.Command("docker", "run", "--rm", "--user", userID, detectImage)
+			output, err := cmd.CombinedOutput()
+			h.AssertNotNil(t, err)
+			expected := "failed to read buildpack order file"
+			h.AssertStringContains(t, string(output), expected)
+		})
+	})
+
+	when("no buildpack group passed detection", func() {
+		it("errors", func() {
+			cmd := exec.Command("docker", "run", "--rm", "--user", userID, "--env", "CNB_ORDER_PATH=/cnb/orders/empty_order.toml", detectImage)
+			output, err := cmd.CombinedOutput()
+			h.AssertNotNil(t, err)
+			expected := "No buildpack groups passed detection."
+			h.AssertStringContains(t, string(output), expected)
+		})
+	})
+
+	when("there is a buildpack group that pass detection", func() {
+		it("writes group.toml and plan.toml", func() {
+			containerName := "test-container-" + h.RandString(10)
+			h.Run(t, exec.Command(
+				"docker",
+				"run",
+				"--name", containerName,
+				"--user", userID,
+				"--env", "CNB_ORDER_PATH=/cnb/orders/simple_order.toml",
+				detectImage))
+			defer h.Run(t, exec.Command("docker", "rm", containerName))
+
+			// check group.toml
+			tempGroupToml, err := ioutil.TempFile("", "")
+			h.AssertNil(t, err)
+			defer os.RemoveAll(tempGroupToml.Name())
+
+			h.Run(t, exec.Command(
+				"docker", "cp",
```

Comment:
Check out `DockerRunAndCopy` which is now in the testhelpers - that could simplify this code.

Reply:
Thans for the advice. I'm using it in my new commit.